### PR TITLE
Issue/6236 – Reports API Reboot

### PR DESCRIPTION
**Pull request by DrewAPicture**
#6236 is the main tracking issue for this aspect of 3DD.

Creating this PR so that we can get a 10,000 ft view of the overall progress the Reports API reboot.

### CODE REVIEW

File: tests/reports/data/tests-endpoint-registry.php [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+		$this->registry->exchangeArray( array() );
+
+		parent::tearDown();
+	}
+
+	/**
+	 * @covers \EDD\Reports\Data\Endpoint_Registry::$item_error_label
+	 */
+	public function test_item_error_label_should_be_reports_endpoint() {
+		$this->assertSame( 'reports endpoint', $this->registry::$item_error_label );
```
Comment (mindctrl) [inline tests/reports/data/tests-endpoint-registry.php]: `$this->registry::$item_error_label` breaks the unit tests, as it's only available on PHP 7+.

File: includes/reports/reports-functions.php [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+ * @param \EDD\Reports\Data\Report $report Report object.
+ */
+function default_display_tables_group( $report ) {
+	if ( $report->has_endpoints( 'tables' ) ) :
+		$tables = $report->get_endpoints( 'tables' );
+		?>
+		<div id="edd-reports-tables-wrap">
+
+			<?php foreach ( $tables as $endpoint_id => $table ) : ?>
+				<h3><?php echo esc_html( $table->get_label() ); ?></h3>
```
Comment (mindctrl) [inline includes/reports/reports-functions.php]: You probably know this already, but `$table` is an array here so this fatals.

Comment (DrewAPicture) [reply includes/reports/reports-functions.php]: `get_endpoints()` returns an array of keyed objects. If `$table` is an array here, then something else is wrong. I think I may have addressed the problem in 8f19622607156b2b83e5a67d3931585856418942.

File: tests/reports/data/tests-endpoint-registry.php (hunk 1 above)
Comment (DrewAPicture) [reply tests/reports/data/tests-endpoint-registry.php]: Out of curiosity, what's the error message? I don't see anything about this that would be PHP 7 specific.

File: includes/reports/reports-functions.php (hunk 2 above)
Comment (mindctrl) [reply includes/reports/reports-functions.php]: Looks like that commit fixed it.

File: tests/reports/data/tests-endpoint-registry.php (hunk 1 above)
Comment (mindctrl) [reply tests/reports/data/tests-endpoint-registry.php]:
```
Parse error: syntax error, unexpected '::' (T_PAAMAYIM_NEKUDOTAYIM) in /home/travis/build/easydigitaldownloads/easy-digital-downloads/tests/reports/data/tests-endpoint-registry.php on line 66
```

File: includes/reports/data/class-endpoint-registry.php [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+		);
+
+		$attributes = array_merge( $defaults, $attributes );
+
+		$attributes['id']    = $endpoint_id;
+		$attributes['views'] = Reports\parse_endpoint_views( $attributes['views'] );
+
+		// Bail if this endpoint ID is already registered.
+		if ( $this->offsetExists( $endpoint_id ) ) {
+			$message = sprintf( 'The \'%1$s\' endpoint already exists and cannont be registered.', $endpoint_id );
```
Comment (mindctrl) [inline includes/reports/data/class-endpoint-registry.php]: Misspelled `cannot` here.

File: includes/reports/class-registry.php [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+	/**
+	 * Retrieves all registered items with a given sorting scheme.
+	 *
+	 * @since 3.0
+	 *
+	 * @param string $sort Optional. How to sort the list of registered items before retrieval.
+	 *                     Accepts 'priority' or 'ID' (alphabetized by item ID), or empty (none).
+	 *                     Default empty.
+	 */
```
Comment (sunnyratilal) [inline includes/reports/class-registry.php]: Missing `@return` tag in PHPDoc.

File: includes/reports/data/charts/v2/class-dataset.php [hunk 5]
```diff
@@ -0,0 +1,318 @@
+<?php
+/**
+ * Reports API - Dataset class
+ *
+ * @package     EDD
+ * @subpackage  Reports\Data\Charts
+ * @copyright   Copyright (c) 2018, Pippin Williamson
+ * @license     http://opensource.org/licenses/gpl-2.0.php GNU Public License
+ * @since       3.0
+ */namespace EDD\Reports\Data\Charts\v2;
```
Comment (sunnyratilal) [inline includes/reports/data/charts/v2/class-dataset.php]: Could this `namespace` be moved on to a separate line please?

**Comment by sunnyratilal** [comment]
@DrewAPicture I've looked over the codebase extensively.  I can't see anything that would constitute as _incorrect_ with any of the methods, or classes.  The unit tests definitely vouch for that.

I've left some class formatting comments but otherwise, this is good to merge. 👍 

Again, truly amazing work done here. Kudos. 💯

**Pull request merged**